File: java/com/countrygamer/arcanacraft/client/gui/GuiArcana.java
package com.countrygamer.arcanacraft.client.gui;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import net.minecraft.client.gui.GuiButton;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.common.IExtendedEntityProperties;

import com.countrygamer.arcanacraft.common.ArcanaCraft;
import com.countrygamer.arcanacraft.common.extended.ExtendedArcanePlayer;
import com.countrygamer.countrygamercore.Base.Plugin.extended.ExtendedEntity;
import com.countrygamer.countrygamercore.Base.client.gui.GuiScreenBase;
import com.countrygamer.countrygamercore.lib.UtilRender;

public class GuiArcana extends GuiScreenBase {
	
	private String currPage;
	private Map<String, ComponentPage> pages = new HashMap<String, ComponentPage>();
	private GuiButton infoTab, quomiTab, skillTab;
	
	public ExtendedArcanePlayer arcanePlayer;
	
	public GuiArcana(EntityPlayer player, boolean revertPages) {
		super("", UtilRender.getResource(ArcanaCraft.pluginID, "gui/", "Scroll"));
		this.xSize = 200;
		this.ySize = 200;
		IExtendedEntityProperties props = ExtendedEntity.getExtended(player,
				ExtendedArcanePlayer.class);
		if (props != null) this.arcanePlayer = (ExtendedArcanePlayer) props;
		if (revertPages) {
			this.currPage = "Info";
		}
		else {
			this.currPage = this.arcanePlayer.getCurrentArcanaPage();
		}
		
		// this.arcanePlayer.printQuoms();
	}
	
	@SuppressWarnings("unchecked")
	@Override
	public void initGui() {
		super.initGui();
		this.createPages();
		this.initPages();
		
		int bID = -1;
		this.buttonList.add(this.infoTab = new GuiButton(bID += 1, this.guiLeft + this.xSize - 10,
				this.guiTop + 5, 20, 20, ""));
		this.buttonList.add(this.quomiTab = new GuiButton(bID += 1, this.guiLeft + this.xSize - 10,
				this.guiTop + 35, 20, 20, ""));
		this.buttonList.add(this.skillTab = new GuiButton(bID += 1, this.guiLeft + this.xSize - 10,
				this.guiTop + 65, 20, 20, ""));
		
		ComponentPageQuomi page = (ComponentPageQuomi) this.pages.get("Quomi");
		page.selected = arcanePlayer.getCurrentSelectedQuomIndex();
		
	}
	
	private void createPages() {
		int a = (this.width - this.xSize) / 2;
		int b = this.xSize;
		int c = (this.height - this.ySize) / 2;
		int d = this.ySize;
		int e = this.grayTextColor;
		
		this.pages.put("Info", new ComponentPage(this, a, b, c, d, e) {
			
			@Override
			public void init() {
				
			}
			
			@Override
			public void drawForeground() {
				String title = "ArcanaCraft; The Arcana";
				mc.fontRenderer.drawString(title, this.guiLeft + (this.xSize / 2)
						- (mc.fontRenderer.getStringWidth(title) / 2), this.guiTop + 5,
						this.grayColor);
				
			}
			
			@Override
			public void drawBackground(int mouseX, int mouseY, float renderPartialTicks) {
				
			}
			
		});
		
		this.pages.put("Quomi", new ComponentPageQuomi(this, a, b, c, d, e, this.arcanePlayer));
		
		this.pages.put("Skill Tree", new ComponentTree(this, a, b, c, d, e, this.arcanePlayer));
		
	}
	
	public void initPages() {
		for (String pageKey : this.pages.keySet()) {
			this.pages.get(pageKey).init();
		}
	}
	
	@Override
	public void updateScreen() {
		super.updateScreen();
		
		if (this.pages.containsKey(this.currPage)) this.pages.get(this.currPage).update();
		
	}
	
	@Override
	protected void actionPerformed(GuiButton gB) {
		if (gB.id == this.infoTab.id) {
			this.currPage = "Info";
		}
		else if (gB.id == this.quomiTab.id) {
			this.currPage = "Quomi";
		}
		else if (gB.id == this.skillTab.id) {
			this.currPage = "Skill Tree";
		}
	}
	
	@Override
	protected void mouseClicked(int x, int y, int mouseButton) {
		super.mouseClicked(x, y, mouseButton);
		if (this.pages.containsKey(this.currPage))
			this.pages.get(this.currPage).onMouseClick(x, y, mouseButton);
	}
	
	@Override
	protected void foregroundText() {
		if (this.pages.containsKey(this.currPage)) {
			this.pages.get(this.currPage).drawForeground();
		}
	}
	
	@Override
	protected void drawGuiContainerBackgroundLayer(float renderPartialTicks, int i, int j) {
		super.drawGuiContainerBackgroundLayer(renderPartialTicks, i, j);
		if (this.pages.containsKey(this.currPage) && !this.currPage.equals("Skill Tree")) {
			this.pages.get(this.currPage).drawBackground(i, j, renderPartialTicks);
		}
	}
	
	final ResourceLocation frameBkgd = UtilRender.getResource(ArcanaCraft.pluginID, "gui/",
			"Scroll_Frame");
	
	@Override
	protected ResourceLocation getBackground() {
		return this.currPage.equals("Skill Tree") ? frameBkgd : super.getBackground();
	}
	
	@Override
	protected void drawBackground(int mouseX, int mouseY, float renderPartialTicks) {
		if (this.pages.containsKey(this.currPage) && this.currPage.equals("Skill Tree")) {
			this.pages.get(this.currPage).drawBackground(mouseX, mouseY, renderPartialTicks);
		}
		super.drawBackground(mouseX, mouseY, renderPartialTicks);
	}
	
	protected void backgroundObjects() {
	}
	
	@Override
	public void addHoverInformation(int mouseX, int mouseY, List<String> hoverInfo) {
		if (this.pages.containsKey(this.currPage))
			this.pages.get(this.currPage).addHoverInformation(mouseX, mouseY, hoverInfo);
	}
	
}

File: src/main/scala/com/countrygamer/arcanacraft/client/gui/GuiArcana.scala
package com.countrygamer.arcanacraft.client.gui

import java.util
import java.util.List

import com.countrygamer.arcanacraft.common.ArcanaCraft
import com.countrygamer.arcanacraft.common.extended.ArcanePlayer
import com.countrygamer.cgo.client.gui.GuiScreenWrapper
import com.countrygamer.cgo.common.lib.util.UtilRender
import com.countrygamer.cgo.wrapper.common.extended.ExtendedEntityHandler
import net.minecraft.client.gui.GuiButton
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.ResourceLocation
import net.minecraftforge.common.IExtendedEntityProperties

/**
 *
 *
 * @author CountryGamer
 */
class GuiArcana(player: EntityPlayer, revertPages: Boolean) extends GuiScreenWrapper(200, 200) {

	final val frameForTree: ResourceLocation =
		UtilRender.getResource(ArcanaCraft.pluginID, "gui/", "Scroll_Frame")

	private var arcanePlayer: ArcanePlayer = null

	private var currentPage: String = "Info"
	private var infoTab, quomiTab, skillTab: GuiButton = null
	private val pages: util.HashMap[String, Page] = new util.HashMap[String, Page]()

	// Default Constructor
	{
		this.setupGui("", UtilRender.getResource(ArcanaCraft.pluginID, "gui/", "Scroll"))

		val props: IExtendedEntityProperties = ExtendedEntityHandler
				.getExtended(player, classOf[ArcanePlayer])
		if (props != null) {
			this.arcanePlayer = props.asInstanceOf[ArcanePlayer]
		}

		if (!revertPages) {
			//this.currentPage = this.arcanePlayer.getCurrentArcanaPage()
		}

	}

	// End Constructor

	override def initGui(): Unit = {
		super.initGui()

		var bID: Int = -1
		bID += 1
		this.infoTab =
				new GuiButton(bID, this.guiLeft + this.xSize - 10, this.guiTop + 5, 20, 20, "")
		this.buttonList.asInstanceOf[List[GuiButton]].add(this.infoTab)
		bID += 1
		this.quomiTab =
				new GuiButton(bID, this.guiLeft + this.xSize - 10, this.guiTop + 35, 20, 20, "")
		this.buttonList.asInstanceOf[List[GuiButton]].add(this.quomiTab)
		bID += 1
		this.skillTab =
				new GuiButton(bID, this.guiLeft + this.xSize - 10, this.guiTop + 65, 20, 20, "")
		this.buttonList.asInstanceOf[List[GuiButton]].add(this.skillTab)

		this.initPages()

	}

	def initPages(): Unit = {

		this.pages.put("Info", new Page(this) {

			override def init(): Unit = {

			}

			override def drawForeground(): Unit = {
				this.drawTitle("ArcanaCraft; The Arcana")
			}

			override def drawBackground(mouseX: Int, mouseY: Int,
					renderPartialTicks: Float): Unit = {

			}

		})

		this.pages.put("Quomi", new PageQuomi(this, this.arcanePlayer))

		this.pages.put("Skill Tree", new PageTree(this, this.arcanePlayer))

		val iterator: util.Iterator[String] = this.pages.keySet().iterator()
		while (iterator.hasNext) {
			this.pages.get(iterator.next()).init()

		}

	}

	override def updateScreen(): Unit = {
		super.updateScreen()

		if (this.pages.containsKey(this.currentPage)) {
			this.pages.get(this.currentPage).onUpdate()

		}

	}

	override def actionPerformed(button: GuiButton): Unit = {
		if (button.id == this.infoTab.id) {
			this.currentPage = "Info"
		}
		else if (button.id == this.quomiTab.id) {
			this.currentPage = "Quomi"
		}
		else if (button.id == this.skillTab.id) {
			this.currentPage = "Skill Tree"
		}

	}

	override def mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int): Unit = {
		super.mouseClicked(mouseX, mouseY, mouseButton)
		if (this.pages.containsKey(this.currentPage)) {
			this.pages.get(this.currentPage).onMouseClick(mouseX, mouseY, mouseButton)
		}

	}

	override protected def drawGuiForegroundLayer(mouseX: Int, mouseY: Int,
			renderPartialTicks: Float): Unit = {
		super.drawGuiForegroundLayer(mouseX, mouseY, renderPartialTicks)

		if (this.pages.containsKey(this.currentPage)) {
			this.pages.get(this.currentPage).drawForeground()
		}
	}

	override protected def drawGuiBackgroundLayer(mouseX: Int, mouseY: Int,
			renderPartialTicks: Float): Unit = {

		if (this.pages.containsKey(this.currentPage) && this.currentPage.equals("Skill Tree")) {
			this.pages.get(this.currentPage).drawBackground(mouseX, mouseY, renderPartialTicks)
		}

		super.drawGuiBackgroundLayer(mouseX, mouseY, renderPartialTicks)

		if (this.pages.containsKey(this.currentPage) && !this.currentPage.equals("Skill Tree")) {
			this.pages.get(this.currentPage).drawBackground(mouseX, mouseY, renderPartialTicks)
		}

	}

	override def getBackgound(): ResourceLocation = {
		if (this.currentPage.equals("Skill Tree"))
			this.frameForTree
		else
			super.getBackgound()
	}

	override protected def addInformationOnHover(mouseX: Int, mouseY: Int,
			renderPartialTicks: Float, hoverInfo: List[String]): Unit = {
		if (this.pages.containsKey(this.currentPage)) {
			this.pages.get(this.currentPage).addHoverInformation(mouseX, mouseY, hoverInfo)
		}

	}

}
